Dynamic ops

samikama opened the pull request:
With this PR, we introduce dynamically generated TRTEngines, that is TRT Engine construction time is postponed to execution time to support graphs where shape inference fails.

samikama committed:
Subgraph to graphdef

samikama committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into dyn_op

samikama committed:
Added segment graphdef conversion functions

Functional Dyn Ops

samikama committed:
Merge with master

samikama committed:
Merge with upstream

samikama committed:
* Use VLOG(1) instead of std::cout in remapper.cc
* Remove op_op_lib dependency from ScopedAllocator. This dependency is
  already satisfied through core and causes a fatal for libraries that
  uses meta_optimizer due to double registration.

samikama committed:
Fix missing return statement

drpngx commented:
Is that related or flaky?

```
147/375 Test #144: T:/src/github/tensorflow/tensorflow/python/kernel_tests/fifo_queue_test.py ..............................................***Failed   24.19 sec
......................................................................FF.........
======================================================================
FAIL: testResetOfBlockingOperation (__main__.FIFOQueueTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "T:/src/github/tensorflow/tensorflow/python/kernel_tests/fifo_queue_test.py", line 1299, in testResetOfBlockingOperation
    t.join()
  File "C:\Python35\lib\site-packages\tensorflow\python\framework\test_util.py", line 1032, in join
    self._testcase.fail("Error in checkedThread: %s" % str(self._exception))
AssertionError: Error in checkedThread: Exception of type <class 'tensorflow.python.framework.errors_impl.CancelledError'>: Session has been closed.
======================================================================
FAIL: testResetOfBlockingOperation (__main__.FIFOQueueTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Python35\lib\site-packages\tensorflow\python\framework\test_util.py", line 738, in tearDown
    thread.check_termination()
  File "C:\Python35\lib\site-packages\tensorflow\python\framework\test_util.py", line 1065, in check_termination
    self._testcase.fail("A checked thread was not joined.")
AssertionError: A checked thread was not joined.
----------------------------------------------------------------------
```

samikama commented:
@drpngx it is very unlikely the error is because of this PR. This module is disabled on Windows builds.

drpngx commented:
Thanks for checking! Let's see what @aaroey has to say.

aaroey commented on tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -30,85 +37,300 @@ using IRuntime = nvinfer1::IRuntime;
+                                                  ->implementation()
+                                                  ->CudaStreamMemberHack()));
+    ah->Ref();  // Increment count for calculating calibration data
+    calib_res->calibrator_->setBatch(input_data, *stream, ah);
```
Since setBatch() is a sync method, can we not passing `ah` but do the ref/unref outside? Actually, why do we need to ref/unref here?

aaroey commented on tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -30,85 +37,300 @@ using IRuntime = nvinfer1::IRuntime;
-  int num_batch = 0;
-  for (int i = 0; i < context->num_inputs(); i++) {
+  int num_batch = ctx->input(0).shape().dim_size(0);
+  int smallest_engine = 0;
```
Please make a new method maybe called 'FindOrCreateEngine' to encapsulate these code from this line to line 267.

aaroey commented on tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -117,51 +339,238 @@ void TRTEngineOp::Compute(OpKernelContext* context) {
+}
+// template <typename T>
+// using destroyed_ptr = std::shared_ptr<T, TRTEngineOp::Destroyer<T>>;
+TRTEngineOp::EngineCtxPair TRTEngineOp::get_engine(int batch_size,
```
GetEngine

aaroey commented on tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -30,85 +37,300 @@ using IRuntime = nvinfer1::IRuntime;
+                 << " failed Running native segment";
+    ExecuteNativeSegment(ctx, ah);
+    return;
+    // ctx->SetStatus(tensorflow::errors::Unavailable(
```
Please remove.

aaroey commented on tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -30,85 +37,300 @@ using IRuntime = nvinfer1::IRuntime;
       num_batch = input_shape.dim_size(0);
       if (num_batch > trt_engine_ptr_->getMaxBatchSize()) {
-        LOG(FATAL) << "input tensor batch larger than max_batch_size: "
+        LOG(ERROR) << "input tensor batch " << num_batch
```
How can this happen?

samikama replied:
It should not be possible. Removed

aaroey commented on tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -117,51 +339,238 @@ void TRTEngineOp::Compute(OpKernelContext* context) {
+                                                   bool ignore_dim_change) {
+  tensorflow::mutex_lock lock(engine_mutex_);
+  if (static_engine) {
+    if (engine_map.size()) {
```
nit: this method is too long, please refactor and put each self-contained logic to a separated method.

samikama replied:
This logic needs to change but it will be done in a subsequent PR. Would you accept a TODO?

aaroey replied:
Sure, thanks.

aaroey commented on tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -117,51 +339,238 @@ void TRTEngineOp::Compute(OpKernelContext* context) {
+// using destroyed_ptr = std::shared_ptr<T, TRTEngineOp::Destroyer<T>>;
+TRTEngineOp::EngineCtxPair TRTEngineOp::get_engine(int batch_size,
+                                                   OpKernelContext* ctx,
+                                                   bool ignore_dim_change) {
```
ignore_dim_change is not used?

samikama replied:
Engine generation logic needs expansion which is left to another PR due to already large size of this PR, where this will be needed

aaroey replied:
Then please add a TODO.

aaroey commented on tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -117,51 +339,238 @@ void TRTEngineOp::Compute(OpKernelContext* context) {
+  }
+}
+
+tensorflow::Status TRTEngineOp::AllocateCalibrationResources(
```
Is this purely copied from TrtCalibOp? Or any diffs?

samikama replied:
CalibOp is not there anymore

aaroey replied:
Yes, I know, I just wanted to make sure the logic is same (so I can stop reviewing it)

samikama replied:
Yes, it is mostly copy-paste with small modifications

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_graph.h:
```diff
@@ -32,6 +32,36 @@ namespace convert {
 
 // This method converts an already generated calibration graph which was used in
 // calibration runs to an inference graph
+struct ConversionParams {
```
Please move the comment above.

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -372,42 +191,418 @@ tensorflow::Status ConvertGraphDefToTensorRT(
+  cp.minimum_segment_size = minimum_segment_size;
+  cp.graph_properties = &static_graph_properties;
+  cp.max_workspace_size_bytes = max_workspace_size_bytes;
+  // return ConvertAfterShapes(gdef, output_names, max_batch_size,
```
Please remove.

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -361,8 +177,11 @@ tensorflow::Status ConvertGraphDefToTensorRT(
   device_properties.mutable_environment()->insert({"architecture", "6"});
-  tensorflow::grappler::Cluster* cluster =
-      new tensorflow::grappler::VirtualCluster({{"/GPU:0", device_properties}});
+  device_properties.set_num_cores(3584);
```
Are these for debugging purposes only?

samikama replied:
Updates to grappler requires num_cores and frequency to be set since they added a check that FATALs if flops calculated using these values is not greater than 0.

aaroey replied:
Would you please add a comment to explain these numbers then?

samikama replied:
Yes they are. Added comments

aaroey reviewed (changes requested):
Thanks for the change! Here are some initial comments, more will come.

Also, have you run all the available tests including tf_trt_integration_test, test_tftrt, and the one in the model repo, to make sure this doesn't break things?

samikama commented on configure.py:
```diff
@@ -943,6 +943,35 @@ def set_tf_cudnn_version(environ_cp):
   write_action_env_to_bazelrc('TF_CUDNN_VERSION', tf_cudnn_version)
 
 
+def is_cuda_compatible(lib, cuda_ver, cudnn_ver):
```
There is a possibility of static linking TRT4.0 with cudnn. This new method doesn't assume it is always linked.

samikama commented on tensorflow/contrib/tensorrt/shape_fn/trt_shfn.cc:
```diff
@@ -31,7 +31,13 @@ namespace shape_inference {
   tensorflow::tensorrt::Logger logger;
   string serialized_engine;
-  TF_RETURN_IF_ERROR(context->GetAttr("serialized_engine", &serialized_engine));
+  if(true){
```
Sorry. missed that

samikama commented on tensorflow/contrib/tensorrt/convert/convert_graph.h:
```diff
@@ -32,6 +32,36 @@ namespace convert {
+  int minimum_segment_size;
+  const tensorflow::grappler::GraphProperties* graph_properties;
+  const tensorflow::grappler::Cluster* cluster;
+  bool is_dyn_op;
```
These are not set through the constructor but filled in later. Setting them const means they have to be  set in construction time.

samikama commented on tensorflow/contrib/tensorrt/resources/trt_int8_calibrator.h:
```diff
@@ -62,6 +68,7 @@ struct TRTInt8Calibrator : public nvinfer1::IInt8EntropyCalibrator {
   bool calib_running_;
   bool batch_is_set_;
   string engine_name_;
+  string calibration_table;
```
ok

samikama commented on tensorflow/contrib/tensorrt/resources/trt_resources.h:
```diff
@@ -46,6 +46,18 @@ class TRTCalibrationResource : public tensorflow::ResourceBase {
   ~TRTCalibrationResource() {
     VLOG(0) << "Destroying Calibration Resource " << std::endl << DebugString();
+    builder_->destroy();
+    builder_ = nullptr;
```
ok

samikama commented on tensorflow/contrib/tensorrt/python/trt_convert.py:
```diff
@@ -141,7 +164,15 @@ def py3string(inp):
+      break
+  if not is_calib_graph:
+    tf_logging.error(
+        "Not a calib graph. Doesn't seem to contain any calibration nodes.")
```
Do we foresee a scenario where graph is composed of int8 and FP32 nodes? Right now all nodes are of the same precision and it is not possible to mix different precision engines.

aaroey replied:
For example, when the trt scope is in place, user may annotate some nodes with int8 and others with fp32.

samikama commented on tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -30,85 +37,300 @@ using IRuntime = nvinfer1::IRuntime;
+    if (max_cached_engines > cached_engine_batches.size()) {
+      smallest_engine = num_batch;
+      cached_engine_batches.push_back(num_batch);
+      std::sort(cached_engine_batches.begin(), cached_engine_batches.end());
```
Yes.

samikama commented:
@aaroey I am getting issues with model tests but these seem to be related with current state of the master. FATALs are happening during execution of native graphs. Our tests pass fine.

samikama committed:
Review changes

samikama committed:
Address review comments and add a check for INT8 engine construction for calibration

samikama committed:
Merge branch 'master' into DynamicOps

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -318,96 +119,503 @@ tensorflow::Status BuildNodeMap(
+  return status;
+}
+
+// Function to construct a funcdef from the segment and add it to the graph.
```
Please move the comment to the header.

samikama replied:
This is an internal method. Not exposed in header.

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -318,96 +119,503 @@ tensorflow::Status BuildNodeMap(
+    VLOG(1) << "Adding " << StrCat(name, "_Arg");
+    node_builder.Attr("T", node->output_type(0)).Attr("index", i).Finalize(&nd);
+    tensorflow::Status s;
+    auto nArg = sgraph.AddNode(nd, &s);
```
Please fix all the names like this.

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -318,96 +119,503 @@ tensorflow::Status BuildNodeMap(
+  }
+
+  for (int i = 0; i < num_inputs; ++i) {
+    auto name = StrCat("InputPH_", i);
```
I don't quite understand: why do the conversion from InputPH_ to _Arg here? Why not directly construct the _Arg nodes from the very beginning?

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -318,96 +119,503 @@ tensorflow::Status BuildNodeMap(
+      VLOG(3) << nd.DebugString();
+    }
+    tensorflow::Status s;
+    auto nRet = sgraph.AddNode(nd, &s);
```
Please fix the name.

samikama commented on tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -117,51 +340,238 @@ void TRTEngineOp::Compute(OpKernelContext* context) {
+      infer->setGpuAllocator(allocator_.get());
+#endif
+      std::shared_ptr<nvinfer1::ICudaEngine> static_engine(
+          infer->deserializeCudaEngine(serialized_segment_.c_str(),
```
That means we will need different fields in the op. In essence it is really a serialized_segment, either in TF language or in TRT language.

samikama commented on tensorflow/contrib/tensorrt/kernels/trt_engine_op.h:
```diff
@@ -31,31 +34,75 @@ limitations under the License.
+      allocators_;
+  string serialized_segment_;
+  string funcdef_name_;
+  string calibration_data_;
```
Done

samikama commented on tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -30,85 +37,306 @@ using IRuntime = nvinfer1::IRuntime;
+  int num_binding = ctx->num_inputs() + ctx->num_outputs();
+  std::vector<void*> buffers(num_binding);
+  int smallest_engine = GetEngineBatch(ctx);
+  if (smallest_engine < 0) return;
```
Error is already set on GetEngineBatch() call

samikama commented on tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -117,51 +345,246 @@ void TRTEngineOp::Compute(OpKernelContext* context) {
+  // TODO(sami): This method needs to be re-written to use resource manager and
+  // with LRU mechanism option.
+  tensorflow::mutex_lock lock(engine_mutex_);
+  if (static_engine_) {
```
No, Static engine is for the cases where everything is fixed and can be determined offline. For such networks there is no point of constructing the engine on the fly. These mostly for final deployment cases.

samikama replied:
There are cases where initialization of the engine is taking in the order or tens of minutes. It is not acceptable to ask every instantiation of an inference network wait for tens of minutes before it can process any input. For such networks, portability is not an issue, there is only one target device. Portability and maintenance argument applies to anything in TensorFlow. If there are backward incompatible changes, user needs to update the graph. Also the flag is there for a reason. There are users which want dynamic construction and there are users which want static construction. Each has pros and cons and users are free to choose whichever is suitable for them. We want users to be able to choose. Static engine is backwards compatible behavior. We may switch the default to dynamic engine but I would prefer doing that gradually, after users have a chance to try and see which one is most suitable for them. In short static construction is a needed feature.

samikama commented on tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -117,51 +345,246 @@ void TRTEngineOp::Compute(OpKernelContext* context) {
+    VLOG(1) << "Starting calibration thread, Calibration Resource @ " << cres;
+    auto s = tensorflow::tensorrt::convert::ConvertSubgraphToEngine(
+        *segment_graph, cres->builder_, shapes, &cres->engine_,
+        tensorflow::tensorrt::convert::INT8MODE);  // will loop until we
```
It triggers calibrator pull, which continues until calibrator returns false;

samikama commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -318,96 +119,503 @@ tensorflow::Status BuildNodeMap(
+  auto trt_rm = tensorflow::tensorrt::TRTResourceManager::instance();
+  auto calib_rm = trt_rm->getManager("TRTCalibration");
+  int num_nodes = infer_graph->node_size();
+  for (int i = 0; i < num_nodes; ++i) {
```
This one is a two-pass approach. In previous case we were updating connections and removing nodes as we add TRTEngineOps thus some connections were getting lost if they are absorbed in other engines. If we start from RTO, connecting edges of the upstream nodes were updated before we absorb them into engines. Here we first collect connections of the engines and then start putting in nodes.

samikama commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -318,96 +119,503 @@ tensorflow::Status BuildNodeMap(
+      if (segment_nodes.count(input_node->name()) == 0) {
+        if (input_node->type_string() ==
+            "Const") {  // Add constant input into segment
+          subgraph_node_ids.push_back(input_node->id());
```
Can you clarify the question? if you are talking about the Const node, we copy every node into a subsegment graph before we remove them. Const will be copied into all segments that are sharing it.

samikama commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -2140,562 +2144,238 @@ void Converter::register_op_converters() {
+  VLOG(1) << "Starting engine conversion ";
+  Converter converter(trt_network.get(), ws.get(), precision_mode == FP16MODE);
+  std::vector<std::pair<string, string>> output_tensors;
+  for (const auto& node_def : gdef.node()) {
```
gdef.node() is already in topological order by construction. subgraph_node_ids is already topo sorted when nodes added into graphdef at ConvertSegmentToGraphDef() call

samikama commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -318,96 +119,503 @@ tensorflow::Status BuildNodeMap(
+      input_shapes.at(conn.port_number) = in_shape;
+      shapes.at(conn.port_number) = conn.outside_shape;
+    }
+    string input_node = conn.outside_node_name;
```
They are updated if other engine contains them as outputs. We can const& them

samikama commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -318,96 +119,503 @@ tensorflow::Status BuildNodeMap(
+    if (!status.ok()) {
+      return status;
+    }
+    if (engine) {
```
No, We are not checking whether engine is non-null in ConvertSubgraphToEngine(). Builder may fail on constructing an engine (for example not enough workspace)

samikama commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -318,96 +119,503 @@ tensorflow::Status BuildNodeMap(
+    return status;
+  }
+  VLOG(1) << "Adding TRTEngine " << info.engine_name << " to graph";
+  engine_node = graph->AddNode(*trtNode, &status);
```
At line 712, if CreateTRTNode() succeeds.

samikama committed:
Further changes for review requests

samikama committed:
Merge conflict fix

samikama commented on tensorflow/contrib/tensorrt/convert/convert_nodes.h:
```diff
@@ -32,12 +32,13 @@ limitations under the License.
 
 namespace tensorflow {
 namespace tensorrt {
+static const string kInputPHName = "InputPH_";
```
Ok. Will do, but I added them as strings since I saw examples of static const string being used elsewhere such as in common_runtime, grappler and framework.

samikama commented on tensorflow/contrib/tensorrt/convert/trt_optimization_pass.cc:
```diff
@@ -204,10 +204,18 @@ tensorflow::Status TRTOptimizationPass::Optimize(
+  cp.minimum_segment_size = minimum_segment_size_;
+  cp.graph_properties = &static_graph_properties;
+  cp.cluster = cluster;
+  cp.is_dyn_op = false;
```
By default we want to construct static ops. There are networks with static shapes that can take tens of minutes to construct. We can't construct such networks in real time.

samikama commented on tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -30,138 +37,548 @@ using IRuntime = nvinfer1::IRuntime;
+    VLOG(1) << "Starting calibration thread, Calibration Resource @ " << cres;
+    auto s = tensorflow::tensorrt::convert::ConvertSubgraphToEngine(
+        *segment_graph, cres->builder_, shapes, &cres->engine_,
+        tensorflow::tensorrt::convert::INT8MODE);  // calibrator will loop until we
```
Isn't this an implementation detail? Do users really need to know that?

samikama committed:
Address review comments and fix some issues

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -692,16 +714,67 @@ tensorflow::Status ConvertAfterShapes(ConversionParams& params) {
+      LOG(WARNING)
+          << "Can't determine the device constructing an allocator at device "
+          << found_device;
+      tensorflow::GPUOptions gpuoptions;
```
Please don't create the allocator here. Remember that allocator is process-wide resource, so the allocator created here will conflict with the ones created by the user explicitly via tf.Session() or any higher level API that will create the session. For now, let's return an error at line 763 or use the TRTCudaAllocator instead, and add a TODO(aaroey), we'll figure out how to find the allocator in this case.

samikama replied:
Are you sure, If user created an allocator on the found device, pm->GetGPUAllocator() will return it and won't construct an allocator. Only problem could be that the if this is instantiated first. It will construct default allocator with allow_growth option enabled. Do we still want that?

aaroey replied:
Yes, if user created an allocator before this will return that one instead of creating a new one. Also, GetGPUAllocator accepts a TfGpuId whose value is **NOT** the device id, so the allocator it returns could be associated with any device, not necessary the one you want here.

aaroey replied:
And, with the cluster problem in OptimizeGraph() fixed, I think we don't need this path any more as we can always get the device.

samikama replied:
>Also, GetGPUAllocator accepts a TfGpuId whose value is NOT the device id, 
I doubt it since device is choosen by querying the GpuIdManager::TfToCudaGpuId(tf_gpu_id, &cuda_gpu_id); and process manager uses the same function to map between tf and cuda gpu ids.

>And, with the cluster problem in OptimizeGraph() fixed, I think we don't need this path any more as we can always get the device.
Is this available right now?

aaroey replied:
The cluster fix should be available with next sync.

samikama replied:
Ok. We can update this with a new PR after it is synced then

samikama commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -417,87 +639,160 @@ tensorflow::Status ConvertAfterShapes(
+    auto pm = tensorflow::ProcessState::singleton();
+    if (params.cluster) {  // get allocator
+      const auto device =
+          params.cluster->GetDeviceSet()->FindDeviceByName(engine.device);
```
Done

samikama commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -121,12 +121,17 @@ tensorflow::Status BuildNodeMap(
   auto calib_rm = trt_rm->getManager("TRTCalibration");
   int num_nodes = infer_graph->node_size();
+  if (!is_dyn_op) {
+    LOG(WARNING) << "Construction of static int8 engine is not implemented "
```
>As the argument of speed only applies to the final infer engine? 

Unfortunately, that is not true.

samikama commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -401,11 +410,15 @@ tensorflow::Status CreateTRTNode(tensorflow::Graph* graph,
     auto status = ConvertSubgraphToEngine(info.segment_graph_def, builder.get(),
                                           shapes, &engine, info.precision_mode);
     if (!status.ok()) {
+      if (engine) engine->destroy();
```
convertSubGraphEngine takes ICudaEngine**, so I don't think it is possible without api change

samikama commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -692,16 +714,67 @@ tensorflow::Status ConvertAfterShapes(ConversionParams& params) {
+    } else {
+      int found_device = 0;
+      bool try_gpu_ids = true;
+      auto checkDeviceId = [](int tfid) -> int {
```
done

aaroey commented on tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -225,7 +224,7 @@ void TRTEngineOp::ExecuteCalibration(tensorflow::OpKernelContext* ctx,
     const auto device_tensor = dev_tensors_.at(i).AccessTensor(ctx);
     CHECK_EQ(t.TotalBytes(),
-             device_tensor->TotalBytes());  // use the tensor so FW keeps it
+             device_tensor->TotalBytes());  // use the tensor so TF keeps it
```
What I meant is AccessTensor() doesn't make TF keeps the tensor. What makes TF keeps the tensor is dev_tensors_ itself.

samikama commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -692,16 +714,67 @@ tensorflow::Status ConvertAfterShapes(ConversionParams& params) {
                 << dev_allocator;
         alloc.reset(new TRTDeviceAllocator(dev_allocator));
       }
+    } else {
```
done

samikama committed:
Address review comments

samikama committed:
Fix Py3 issue and device placement

samikama committed:
Workaround Grappler funcdef optimization issue

samikama committed:
Fix line too long error on method doc

samikama committed:
Another linter fix

aaroey committed:
Multiple changes:
1. use unique_ptr instead of shared_ptr, and fix a bug in destructor of TrtEngineOp where it did't reset the shared_ptr but a copy of it
2. fix the include order
3. shorten the reference to tensorflow::tensorrt::xxx
4. remove some code that sets something which will be overwritten later
5. fix format, including: function signature, variable names, const reference, etc
6. remove some deadcode
7. add a lot of comments and TODOs
8. in TrtEngineOp, replace the map of allocators with a single unique_ptr
9. in TrtEngineOp, remove parameter ignore_dim_change from GetEngine(), since it always uses member fixed_input_size_

aaroey committed:
Add missing utils.h

aaroey committed:
Use TrtUniquePtrType for all builder/network/engine construction; add build rules for utils.h; add more TODOs

samikama commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -2140,559 +2129,247 @@ void Converter::register_op_converters() {
-      StrAppend(&node_names, node_name, ", ");
-    }
-    VLOG(2) << node_names;
+tensorflow::Status ConvertSubGraphDefToEngine(
```
@aaroey This method converts a graphdef to TRT Engine. It is a full fledged graphdef not subgraphdef. I think naming should be reverted

aaroey replied:
Done

samikama commented on tensorflow/contrib/tensorrt/convert/convert_nodes.h:
```diff
@@ -36,11 +38,13 @@ static const char* kInputPHName = "InputPH_";
-struct EngineConnections {
-  EngineConnections(const string& outside, int out_id, int out_port,
+
+struct EngineConnection {
```
@aaroey  This structure represent connections of an engine, I believe it should be plural

samikama committed:
Address some comments

aaroey committed:
Add some more comments and fix some TODOs

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -318,186 +123,694 @@ tensorflow::Status BuildNodeMap(
-  tensorflow::tensorrt::segment::SegmentOptions segment_options;
+    const std::set<string>& segment_nodes,
+    const std::unordered_map<string, tensorflow::Node*>& node_map,
+    const std::vector<tensorflow::Node*>& reverse_topo_order) {
```
As discussed, this is really reverse_topo_order.

aaroey committed:
Fix name of ConvertSubGraphDefToEngine()

aaroey committed:
Move the builder creation logic into ConvertGraphDefToEngine(), use unique_ptr for TRTCalibrationResource, and fix comments

aaroey committed:
Fix some formatting issues

aaroey committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into DynamicOps

aaroey committed:
Fix internal build errors.
